// aspect.js/src/join_points/method_call.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var join_point_1 = require("../core/join_point");
var pointcut_1 = require("../core/pointcut");
var aspect_1 = require("../core/aspect");
var preconditions_1 = require("./preconditions");
var BLACK_LIST = ['constructor'];
var MethodCallJoinPoint = /** @class */ (function (_super) {
    __extends(MethodCallJoinPoint, _super);
    function MethodCallJoinPoint() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MethodCallJoinPoint.prototype.getTarget = function (fn) {
        return fn.prototype;
    };
    MethodCallJoinPoint.prototype.match = function (target) {
        var _this = this;
        var keys = Object.getOwnPropertyNames(target.prototype);
        keys = keys.filter(function (key) {
            return BLACK_LIST.indexOf(key) < 0;
        });
        var res = keys
            .map(function (key) {
            var descriptor = Object.getOwnPropertyDescriptor(target.prototype, key);
            if (_this.precondition.assert({
                classDefinition: target,
                methodName: key
            }) &&
                typeof descriptor.value === 'function') {
                return key;
            }
            return false;
        })
            .filter(function (val) { return !!val; });
        return res;
    };
    MethodCallJoinPoint.prototype.wrapTarget = function (proto, key, advice, advisedMetadata) {
        var className = proto.constructor.name;
        var bak = proto[key];
        var self = this;
        proto[key] = function () {
            var metadata = self.getMetadata(className, key, bak, arguments, this, advisedMetadata);
            return advice.apply(bak, metadata);
        };
        proto[key].__woven__ = true;
    };
    return MethodCallJoinPoint;
}(join_point_1.JoinPoint));
exports.MethodCallJoinPoint = MethodCallJoinPoint;
function makeMethodCallAdviceDecorator(constr) {
    return function () {
        var selectors = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            selectors[_i] = arguments[_i];
        }
        return function (target, prop, descriptor) {
            var joinpoints = selectors.map(function (selector) {
                return new MethodCallJoinPoint(new preconditions_1.MethodPrecondition(selector));
            });
            var pointcut = new pointcut_1.Pointcut();
            pointcut.advice = new constr(target, descriptor.value);
            pointcut.joinPoints = joinpoints;
            var aspectName = target.constructor.name;
            var aspect = aspect_1.AspectRegistry.get(aspectName) || new aspect_1.Aspect();
            aspect.pointcuts.push(pointcut);
            aspect_1.AspectRegistry.set(aspectName, aspect);
            // For lazy loading
            aspect_1.Targets.forEach(function (_a) {
                var target = _a.target, config = _a.config;
                return aspect.apply(target, config);
            });
            return target;
        };
    };
}
exports.makeMethodCallAdviceDecorator = makeMethodCallAdviceDecorator;
/**
 * Kept for backward compability only.
 * Use {@link MethodCallJoinPoint} instead.
 *
 * @deprecated renamed to MethodCallJoinPoint
 * @see MethodCallJoinPoint
 */
var MethodCallJointPoint = /** @class */ (function (_super) {
    __extends(MethodCallJointPoint, _super);
    function MethodCallJointPoint(precondition) {
        return _super.call(this, precondition) || this;
    }
    return MethodCallJointPoint;
}(MethodCallJoinPoint));
exports.MethodCallJointPoint = MethodCallJointPoint;
