// aspect.js/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var advices_1 = require("./src/advices");
var join_points_1 = require("./src/join_points");
var join_points_2 = require("./src/join_points");
var core_1 = require("./src/core");
exports.Advised = core_1.Advised;
exports.Metadata = core_1.Metadata;
exports.MethodMetadata = core_1.MethodMetadata;
exports._AspectRegistry = core_1.AspectRegistry;
exports._Targets = core_1.Targets;
var join_points_3 = require("./src/join_points");
exports.MemberPrecondition = join_points_3.MemberPrecondition;
exports.beforeMethod = join_points_1.makeMethodCallAdviceDecorator(advices_1.BeforeAdvice);
exports.afterMethod = join_points_1.makeMethodCallAdviceDecorator(advices_1.AfterAdvice);
exports.aroundMethod = join_points_1.makeMethodCallAdviceDecorator(advices_1.AroundAdvice);
exports.onThrowOfMethod = join_points_1.makeMethodCallAdviceDecorator(advices_1.OnThrowAdvice);
exports.beforeSetter = join_points_2.makeFieldSetAdviceDecorator(advices_1.BeforeAdvice);
exports.afterSetter = join_points_2.makeFieldSetAdviceDecorator(advices_1.AfterAdvice);
exports.aroundSetter = join_points_2.makeFieldSetAdviceDecorator(advices_1.AroundAdvice);
exports.onThrowOfSetter = join_points_2.makeFieldSetAdviceDecorator(advices_1.OnThrowAdvice);
exports.beforeGetter = join_points_2.makeFieldGetAdviceDecorator(advices_1.BeforeAdvice);
exports.afterGetter = join_points_2.makeFieldGetAdviceDecorator(advices_1.AfterAdvice);
exports.aroundGetter = join_points_2.makeFieldGetAdviceDecorator(advices_1.AroundAdvice);
exports.onThrowOfGetter = join_points_2.makeFieldGetAdviceDecorator(advices_1.OnThrowAdvice);
exports.beforeStaticMethod = join_points_1.makeStaticMethodAdviceDecorator(advices_1.BeforeAdvice);
exports.afterStaticMethod = join_points_1.makeStaticMethodAdviceDecorator(advices_1.AfterAdvice);
exports.aroundStaticMethod = join_points_1.makeStaticMethodAdviceDecorator(advices_1.AroundAdvice);
exports.onThrowOfStaticMethod = join_points_1.makeStaticMethodAdviceDecorator(advices_1.OnThrowAdvice);
// Async
exports.asyncOnThrowOfMethod = join_points_1.makeMethodCallAdviceDecorator(advices_1.AsyncOnThrowAdvice);
exports.asyncOnThrowOfSetter = join_points_2.makeFieldSetAdviceDecorator(advices_1.AsyncOnThrowAdvice);
exports.asyncOnThrowOfGetter = join_points_2.makeFieldGetAdviceDecorator(advices_1.AsyncOnThrowAdvice);
exports.asyncOnThrowOfStaticMethod = join_points_1.makeStaticMethodAdviceDecorator(advices_1.AsyncOnThrowAdvice);
